lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  // All the edges to the header need to be rewired.
+  for (SILBasicBlock *headerPred : header->getPredecessorBlocks()) {
+    edgesToFix.emplace_back(headerPred, header);
+  }
+
+  for (const SILBasicBlock *bb : loop->getBlocks()) {
+    // Save the values that are escaping this loop in escapingValues set.
+    auto saveEscaping = [this](SILValue value) {
+      for (const auto *use : value->getUses()) {
+        const SILInstruction *useInst = use->getUser();
+        if (!this->loop->contains(useInst->getParent())) {
+          escapingValues.insert(value);
+          break;
+        }
+      }
+    };
+    if (bb != header) {
+      llvm::for_each(bb->getArguments(), saveEscaping);
+    }
+    for (const SILInstruction &inst : *bb) {
+      llvm::for_each(inst.getResults(), saveEscaping);
+    }
+  }
+}
+
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst *SingleExitLoopTransformer::appendArguments(
+    TermInst *termInst, SILBasicBlock *target, ArrayRef<SILValue> newArgs) {
+  SILBuilder builder(termInst);
+  TermInst *newTermInst = nullptr;
+
+  auto createArgsArray = [target,
+                          &newArgs](const OperandValueArrayRef &currentArgs,
+                                    bool appendNewArgs) {
+    SmallVector<SILValue, 8> args(currentArgs.begin(), currentArgs.end());
+    if (appendNewArgs) {
+      args.append(newArgs.begin(), newArgs.end());
+      assert(args.size() == target->getNumArguments() &&
+             "Number of final arguments does not match target's arguments.");
+    }
+    return args;
+  };
+  if (auto *branch = dyn_cast<BranchInst>(termInst)) {
+    assert(branch->getDestBB() == target &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> args =
+        createArgsArray(branch->getArgs(), /*appendNewArgs=*/true);
+    newTermInst =
+        builder.createBranch(branch->getLoc(), branch->getDestBB(), args);
+  } else if (auto *condBranch = dyn_cast<CondBranchInst>(termInst)) {
+    bool isTrueEdge = condBranch->getTrueBB() == target;
+    assert(((isTrueEdge && condBranch->getTrueBB() == target) ||
+            (!isTrueEdge && condBranch->getFalseBB() == target)) &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> trueArgs = createArgsArray(
+        condBranch->getTrueArgs(), /*appendNewArgs=*/isTrueEdge);
+    SmallVector<SILValue, 8> falseArgs = createArgsArray(
+        condBranch->getFalseArgs(), /*appendNewArgs=*/!isTrueEdge);
+    newTermInst = builder.createCondBranch(
+        condBranch->getLoc(), condBranch->getCondition(),
+        condBranch->getTrueBB(), trueArgs, condBranch->getFalseBB(), falseArgs,
+        condBranch->getTrueBBCount(), condBranch->getFalseBBCount());
+  } else {
+    // At the moment we can only add arguments to br and cond_br.
+    llvm_unreachable("Can't add argument to terminator");
```

Comment:
can we do assert on CondBr on the previous `else if` case, and remove this `else` case?

Reply:
It makes it clear that we only have two possible cases. With this extra else body, it's not clear if we are missing some impl code, to be added in the future. Also the code gets longer.

Reply:
Fair enough. Done. :)